Using a function imported from a host module as a start function causes a panic

**Describe the bug**
If a module defines a start function using an imported function, and the imported function happens to come from a host module a panic will occur.

```
panic: runtime error: index out of range [0] with length 0 [recovered, repanicked]

goroutine 25 [running]:
testing.tRunner.func1.2({0x102c85260, 0x1400001a120})
        /xxx/go/1.25.5/libexec/src/testing/testing.go:1872 +0x190
testing.tRunner.func1()
        /xxx/go/1.25.5/libexec/src/testing/testing.go:1875 +0x31c
panic({0x102c85260?, 0x1400001a120?})
        /xxx/go/1.25.5/libexec/src/runtime/panic.go:783 +0x120
github.com/tetratelabs/wazero/internal/engine/wazevo.(*moduleEngine).NewFunction(0x1400017e780, 0x0)
        /xxx/wazero/internal/engine/wazevo/module_engine.go:180 +0x2ac
github.com/tetratelabs/wazero/internal/engine/wazevo.(*moduleEngine).NewFunction(0x1400030b200?, 0x2b9467c?)
        /xxx/wazero/internal/engine/wazevo/module_engine.go:161 +0x54
github.com/tetratelabs/wazero/internal/wasm.(*ModuleInstance).ExportedFunction(0x1400030b200, {0x102b9467c?, 0x102f0d9a0?})
        /xxx/wazero/internal/wasm/module_instance.go:210 +0x58
github.com/tetratelabs/wazero.(*runtime).InstantiateModule(0x14000455aa0, {0x102cae130, 0x1400011ea20}, {0x102cb00a0, 0x14000013900}, {0x102cb4b00, 0x140000abd98})
        /xxx/wazero/runtime.go:358 +0x2ec
github.com/tetratelabs/wazero.(*runtime).InstantiateWithConfig(0x14000455aa0, {0x102cae130, 0x1400011ea20}, {0x1400001c5c0?, 0x102cb00a0?, 0x140000137c0?}, {0x102cb4b00, 0x140000abd98})
        /xxx/wazero/runtime.go:302 +0x70
github.com/tetratelabs/wazero.(*runtime).Instantiate(0x14000455aa0, {0x102cae130, 0x1400011ea20}, {0x1400001c5c0, 0x2a, 0x40})
        /xxx/wazero/runtime.go:293 +0xa8
github.com/tetratelabs/wazero.TestRuntime_Instantiate_ErrorOnStart.func1(0x14000003c00)
        /xxx/wazero/runtime_test.go:429 +0x16c
testing.tRunner(0x14000003c00, 0x1400037e340)
        /xxx/go/1.25.5/libexec/src/testing/testing.go:1934 +0xc8
created by testing.(*T).Run in goroutine 24
        /xxx/go/1.25.5/libexec/src/testing/testing.go:1997 +0x364
```

**To Reproduce**
Basically, this existing test would reproduce the problem if `wazero` implemented loading wasm text format: https://github.com/wazero/wazero/blob/main/runtime_test.go#L387

The test actually just fails before doing anything because the wasm modules are defined in text format, and hence parsing fails very early on in the process. A corrected version of the test reproduces the problem.

**Expected behavior**
The start function is invoked without error.

**Environment (please complete the relevant information):**
- Go version: 1.25.5
- wazero Version: f7cf1693f3d4c7ffd38d1850028270d1ea891d45
- Host architecture: arm64
- Runtime mode: compiler

--- a/internal/engine/wazevo/module_engine.go
+++ b/internal/engine/wazevo/module_engine.go
@@ -1,13 +1,17 @@
 package wazevo
 
 import (
+	"context"
 	"encoding/binary"
+	"fmt"
 	"unsafe"
 
 	"github.com/tetratelabs/wazero/api"
 	"github.com/tetratelabs/wazero/experimental"
 	"github.com/tetratelabs/wazero/internal/engine/wazevo/wazevoapi"
+	"github.com/tetratelabs/wazero/internal/internalapi"
 	"github.com/tetratelabs/wazero/internal/wasm"
+	"github.com/tetratelabs/wazero/internal/wasmdebug"
 	"github.com/tetratelabs/wazero/internal/wasmruntime"
 )
 
@@ -158,6 +162,21 @@ func (m *moduleEngine) NewFunction(index wasm.Index) api.Function {
 		return imported.me.NewFunction(imported.indexInModule)
 	} else {
 		localIndex -= importedFnCount
+	}
+
+	if source := m.module.Source; source.IsHostModule {
+		// For host modules, we need to look up the GoFunction from the CodeSection.
+		def := source.FunctionDefinition(localIndex)
+		goF := source.CodeSection[localIndex].GoFunc
+		switch typed := goF.(type) {
+		case api.GoFunction:
+			// GoFunction doesn't need looked up module.
+			return &hostFunction{def: def, g: goFunctionAsGoModuleFunction(typed)}
+		case api.GoModuleFunction:
+			return &hostFunction{def: def, lookedUpModule: m.module, g: typed}
+		default:
+			panic(fmt.Sprintf("unexpected GoFunc type: %T", goF))
+		}
 	}
 
 	src := m.module.Source
@@ -327,4 +346,46 @@ func (m *moduleEngine) LookupFunction(t *wasm.TableInstance, typeId wasm.Functio
 
 func moduleInstanceFromOpaquePtr(ptr *byte) *wasm.ModuleInstance {
 	return *(**wasm.ModuleInstance)(unsafe.Pointer(ptr))
+}
+
+type hostFunction struct {
+	internalapi.WazeroOnly
+	def            *wasm.FunctionDefinition
+	lookedUpModule *wasm.ModuleInstance
+	g              api.GoModuleFunction
+}
+
+// goFunctionAsGoModuleFunction converts api.GoFunction to api.GoModuleFunction which ignores the api.Module argument.
+func goFunctionAsGoModuleFunction(g api.GoFunction) api.GoModuleFunction {
+	return api.GoModuleFunc(func(ctx context.Context, _ api.Module, stack []uint64) {
+		g.Call(ctx, stack)
+	})
+}
+
+// Definition implements api.Function.
+func (f *hostFunction) Definition() api.FunctionDefinition { return f.def }
+
+// Call implements api.Function.
+func (f *hostFunction) Call(ctx context.Context, params ...uint64) ([]uint64, error) {
+	typ := f.def.Functype
+	stackSize := typ.ParamNumInUint64
+	rn := typ.ResultNumInUint64
+	if rn > stackSize {
+		stackSize = rn
+	}
+	stack := make([]uint64, stackSize)
+	copy(stack, params)
+	return stack[:rn], f.CallWithStack(ctx, stack)
+}
+
+// CallWithStack implements api.Function.
+func (f *hostFunction) CallWithStack(ctx context.Context, stack []uint64) (err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			builder := wasmdebug.NewErrorBuilder()
+			err = builder.FromRecovered(r)
+		}
+	}()
+	f.g.Call(ctx, f.lookedUpModule, stack)
+	return nil
 }

--- a/runtime_test.go
+++ b/runtime_test.go
@@ -385,22 +385,26 @@ func TestRuntime_Instantiate_DoesntEnforce_Start(t *testing.T) {
 }
 
 func TestRuntime_Instantiate_ErrorOnStart(t *testing.T) {
+	var indexZero wasm.Index
 	tests := []struct {
-		name, wasm string
+		name string
+		wasm *wasm.Module
 	}{
 		{
 			name: "_start function",
-			wasm: `(module
-	(import "" "start" (func $start))
-	(export "_start" (func $start))
-)`,
+			wasm: &wasm.Module{
+				TypeSection:   []wasm.FunctionType{{}},
+				ImportSection: []wasm.Import{{Module: "host", Name: "start", Type: wasm.ExternTypeFunc, DescFunc: 0}},
+				ExportSection: []wasm.Export{{Name: "_start", Type: wasm.ExternTypeFunc, Index: 0}},
+			},
 		},
 		{
 			name: ".start function",
-			wasm: `(module
-	(import "" "start" (func $start))
-	(start $start)
-)`,
+			wasm: &wasm.Module{
+				TypeSection:   []wasm.FunctionType{{}},
+				ImportSection: []wasm.Import{{Module: "host", Name: "start", Type: wasm.ExternTypeFunc, DescFunc: 0}},
+				StartSection:  &indexZero,
+			},
 		},
 	}
 
@@ -420,9 +424,11 @@ func TestRuntime_Instantiate_ErrorOnStart(t *testing.T) {
 				Instantiate(testCtx)
 			require.NoError(t, err)
 
-			// Start the module as a WASI command. We expect it to fail.
-			_, err = r.Instantiate(testCtx, []byte(tc.wasm))
+			// Start the module as a WASI command. We expect it to fail with the error from the
+			// start function.
+			_, err = r.Instantiate(testCtx, binaryencoding.EncodeModule(tc.wasm))
 			require.Error(t, err)
+			require.Contains(t, err.Error(), "ice cream")
 
 			// Close the imported module, which should remove its compiler cache.
 			require.NoError(t, host.Close(testCtx))

